teaclave_attestation/src/quote.rs

```diff
@@ -228,136 +232,139 @@ pub struct SgxQuote {
-    let (attn_report_raw, sig, sig_cert_dec) = mayfail! {
-        attn_report_raw =<< iter.next();
-        sig_raw =<< iter.next();
-        sig =<< base64::decode(&sig_raw);
-        sig_cert_raw =<< iter.next();
-        sig_cert_dec =<< base64::decode_config(&sig_cert_raw, base64::STANDARD);
-        ret (attn_report_raw, sig, sig_cert_dec)
-    }
-    .map_err(|_| CertVerificationError::InvalidCertFormat)?;
-
-    let sig_cert = webpki::EndEntityCert::from(&sig_cert_dec)
-        .map_err(|_| CertVerificationError::InvalidCertFormat)?;
-
-    // Verify if the signing cert is issued by Intel CA
-    let ias_report_ca = BUILD_CONFIG.ias_root_ca_cert;
-    let mut ias_ca_stripped = ias_report_ca.to_vec();
-    ias_ca_stripped.retain(|&x| x != 0x0d && x != 0x0a);
-    let head_len = "-----BEGIN CERTIFICATE-----".len();
-    let tail_len = "-----END CERTIFICATE-----".len();
-    let full_len = ias_ca_stripped.len();
-    let ias_ca_core: &[u8] = &ias_ca_stripped[head_len..full_len - tail_len];
-    let ias_cert_dec = base64::decode_config(ias_ca_core, base64::STANDARD)
-        .map_err(|_| CertVerificationError::InvalidCertFormat)?;
-
-    let mut ca_reader = BufReader::new(&ias_report_ca[..]);
-
-    let mut root_store = rustls::RootCertStore::empty();
-    root_store
-        .add_pem_file(&mut ca_reader)
-        .expect("Failed to add CA");
-
-    let trust_anchors: Vec<webpki::TrustAnchor> = root_store
-        .roots
-        .iter()
-        .map(|cert| cert.to_trust_anchor())
-        .collect();
-
-    let chain: Vec<&[u8]> = vec![&ias_cert_dec];
-
-    let now_func = webpki::Time::try_from(SystemTime::now())
-        .map_err(|_| CertVerificationError::WebpkiFailure)?;
-
-    sig_cert
-        .verify_is_valid_tls_server_cert(
-            SUPPORTED_SIG_ALGS,
-            &webpki::TLSServerTrustAnchors(&trust_anchors),
-            &chain,
-            now_func,
-        )
-        .map_err(|_| CertVerificationError::WebpkiFailure)?;
-
-    // Verify the signature against the signing cert
-    sig_cert
-        .verify_signature(&webpki::RSA_PKCS1_2048_8192_SHA256, &attn_report_raw, &sig)
-        .map_err(|_| CertVerificationError::WebpkiFailure)?;
-
-    // Verify attestation report
-    let attn_report: Value = serde_json::from_slice(attn_report_raw)
-        .map_err(|_| CertVerificationError::BadAttnReport)?;
-
-    // 1. Check timestamp is within 24H (90day is recommended by Intel)
-    let quote_freshness = mayfail! {
-        time =<< attn_report["timestamp"].as_str();
-        let time_fixed = String::from(time) + "+0000";
-        date_time =<< DateTime::parse_from_str(&time_fixed, "%Y-%m-%dT%H:%M:%S%.f%z");
-        let ts = date_time.naive_utc();
-        let now = DateTime::<chrono::offset::Utc>::from(SystemTime::now()).naive_utc();
-        secs =<< u64::try_from((now - ts).num_seconds());
-        ret secs
-    }
-    .map_err(|_| CertVerificationError::BadAttnReport)?;
+        let raw_pub_k = pub_k.to_bytes();
```

Comment:
Here we are verifying the attestation report generated in ‘pub struct IasReport’. It’s better to export a verification function from that struct, making the generation and verification logic in the same module.

Reply:
The `IasReport` is a string of report for server, the client will open the string into `SgxQuote` and `SgxQuoteBody`. I didn't find a good place for this.

Reply:
More specifically, the public key format(reversed bytes without a leading 4u8) in the ias report is defined in IasReport::create_report, it’s better to define the verification logic around that piece of code.
In addition, the comment “ // In the case of MesaTEE, we only allow the uncompressed form.” might be a little confusing. The raw_pub_k is in compressed form because rustls->ring only supports the compressed form.

Reply:
I don't understand. The meanings of these two variables are:

- The `raw_pub_k` is from TLS cert (before extension) by `Secp256k1KeyPair::create_cert_with_extension`.
- The `quote_body.report_body.report_data` is from IAS report (in TLS cert extension) created by `IasReport::create_report`.

This code is to compare `raw_pub_k` with `quote_body.report_body.report_data`. The only difference is there is a first prefix byte (`4u8`) when constructing TBS cert. So, the only thing is to compare the two keys, no need to reverse and others.

Reply:
> I don't understand. The meanings of these two variables are:
> 
> * The `raw_pub_k` is from TLS cert (before extension) by `Secp256k1KeyPair::create_cert_with_extension`.
> * The `quote_body.report_body.report_data` is from IAS report (in TLS cert extension) created by `IasReport::create_report`.
> 
> This code is to compare `raw_pub_k` with `quote_body.report_body.report_data`. The only difference is there is a first prefix byte (`4u8`) when constructing TBS cert. So, the only thing is to compare the two keys, no need to reverse and others.

The IasReport is responsible for the comparison to verify the report data field, not  the caller.

Reply:
In short, we can introduce a separate file to define make_report_data and verify_report_data.

Reply:
Ok, this can be done in the next round of polishing. Let me first move to RPC and come back later.